src/io_ops/mod.rs

```diff
@@ -0,0 +1,103 @@
+use libc::{self, c_void, size_t};
+use log::error;
+
+use std::os::unix::prelude::AsRawFd;
+use std::{
+    ops::Deref,
+    os::unix::io::RawFd,
+    sync::{atomic::AtomicBool, Arc},
+};
+
+use std::io;
+
+#[derive(Debug, Clone)]
+pub struct ArcSubChannel(pub(crate) Arc<SubChannel>);
+
+impl ArcSubChannel {
+    pub fn as_raw_fd(&self) -> &FileDescriptorRawHandle {
+        self.0.as_ref().as_raw_fd()
+    }
+}
+
+impl Deref for ArcSubChannel {
+    type Target = SubChannel;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref()
+    }
+}
+#[async_trait::async_trait]
+impl crate::reply::ReplySender for ArcSubChannel {
+    async fn send(&self, data: &[&[u8]]) {
+        if let Err(err) = SubChannel::send(self.0.as_ref(), data).await {
+            error!("Failed to send FUSE reply: {}", err);
+        }
+    }
+}
+
+/// In the latest version of rust this isn't required since RawFd implements AsRawFD
+/// but until pretty recently that didn't work. So including this wrapper is cheap and allows
+/// us better compatibility.
+#[derive(Debug)]
+pub struct FileDescriptorRawHandle {
+    pub(in crate) fd: RawFd,
+    is_closed: AtomicBool,
+}
+
+impl FileDescriptorRawHandle {
+    pub fn new(fd: RawFd) -> Self {
+        Self {
+            fd,
+            is_closed: AtomicBool::default(),
+        }
+    }
+    pub fn close(&self) {
+        let already_closed = self
+            .is_closed
+            .swap(true, std::sync::atomic::Ordering::SeqCst);
```

Comment:
The atomic here guarantees that a file is only closed once, but there are no guarantees that read or write aren't called on the file after it has been closed.  This is unsafe as FDs are reused.

Reply:
`unsafe` in rust's definition isn't generally taken to include this type of logic behavior anymore/generally afaik. 

nomicon: https://doc.rust-lang.org/nomicon/what-unsafe-does.html

&& similar types in say tokio are not marked as unsafe, e.g. https://docs.rs/tokio/1.2.0/tokio/io/unix/struct.AsyncFd.html
(the new here based on the logic above should be unsafe(though it was discussed in their api, and they decided not to), since there's no guarantee the FD exists, isn't duplicated or anything else). 

this stuff we should limit how much it can go wrong(try make it as hard/impossible as we can) and put the right privacy controls in place, but from what i can work out in rust these days the unsafe keyword is more kept for memory safety.

Reply:
(if you have access to a more concrete guide i'd love to see it, does seem to be lots of confusion in this space in general that i've seen on different crates.)

Reply:
> similar types in say tokio are not marked as unsafe, e.g. https://docs.rs/tokio/1.2.0/tokio/io/unix/struct.AsyncFd.html

Note: `AsyncFd<T>` is not the same. In that case of `AsyncFd<File>` The `AsyncFd` owns the `File` and the file owns the underlying FD and is responsible closing it. In the case of `AsyncFd<RawFd>` the `AsyncFd` owns the `RawFd`, but the `RawFd` is really only a (non-owning) reference to the underlying fd, and is not responsible for closing it. Indeed `RawFd` does nothing on `drop`, it's just a typedef for `c_int`.

This is the [reason that FromRawFd::from_raw_fd is unsafe](https://doc.rust-lang.org/std/os/unix/io/trait.FromRawFd.html#tymethod.from_raw_fd):

> This function is also unsafe as the primitives currently returned have the contract that they are the sole owner of the file descriptor they are wrapping. Usage of this function could accidentally allow violating this contract which can cause memory unsafety in code that relies on it being true.

`AsyncFd<RawFd>` will call `poll` on the underlying FD, but this should be safe as poll *shouldn't* have side-effects. Calling `read` or `write` or `close` do have side-effects, so are unsafe.

Reply:
> unsafe in rust's definition isn't generally taken to include this type of logic behavior anymore/generally afaik. ...these days the unsafe keyword is more kept for memory safety.

To be clear: this isn't unsafe because it itself violates memory safety, but because it could cause memory unsafety in other code that does expect to be the sole owner of a file.

Reply:
Does reading different data than expected by an interleaving seek count as memory in safety? That’s the main sort of memory issue I could think of. But do you have something else in mind ?

Reply:
Memory mapping a file via the fd? Just curious here as to how these interact.

Reply:
https://users.rust-lang.org/t/why-is-fromrawfd-unsafe/39670/5 Seems to suggest this is an historical artifact from an older definition of unsafe ?